src/sentry/search/utils.py

```diff
@@ -595,19 +621,44 @@ def split_query_into_tokens(query):
                 if quote_enclosed:
                     quote_type = char
         if quote_enclosed and char == "\\" and next_char == quote_type:
-            token += next_char
-            idx += 1
+            if next_char is not None:
+                token += next_char
+                idx += 1
         idx += 1
     if not token.isspace():
         tokens.append(token.strip(" "))
     return tokens
 
 
-def parse_query(projects, query, user, environments):
+class ParsedQueryValues(TypedDict, total=False):
+    tags: dict[str, Union[str, list[str], Any]]
+    query: str
+    unassigned: NotRequired[bool]
+    for_review: NotRequired[bool]
+    linked: NotRequired[bool]
+    status: NotRequired[int]
+    assigned_to: NotRequired[Optional[Union[User, Team]]]
+    assigned_or_suggested: NotRequired[Optional[Union[User, Team]]]
+    bookmarked_by: NotRequired[User]
+    subscribed_by: NotRequired[User]
+    first_release: NotRequired[Sequence[str]]
+
+    age_from: NotRequired[Union[datetime, bool]]
+    age_to: NotRequired[Union[datetime, bool]]
+    last_seen_from: NotRequired[Union[datetime, bool]]
+    last_seen_to: NotRequired[Union[datetime, bool]]
+    date_from: NotRequired[Union[datetime, bool]]
+    date_to: NotRequired[Union[datetime, bool]]
+    times_seen: NotRequired[Union[int, float]]
+
+
+def parse_query(
+    projects: Sequence[Project], query: str, user: User, environments: Sequence[Environment]
+) -> ParsedQueryValues:
     # TODO(dcramer): handle query being wrapped in quotes
     tokens = tokenize_query(query)
 
-    results = {"tags": {}, "query": []}
+    results: dict[str, Any] = {"tags": {}, "query": []}
```

Comment:
Hmm, spent some time trying to figure this out but there seems to be some type-checking incompatibilities with `dict` and `TypedDict`. Apparently you can't call `update` on a TypedDict since its supposed to be immutable. 

But then we go ahead and mutate the parsed query values here:
https://github.com/getsentry/sentry/blob/74a5f38809ed265fd738d03209a5b7f7c9971b0f/src/sentry/api/endpoints/group_events.py#L126-L127

Looks like my usage of `TypedDict` here is incorrect. Hmm. I think I'll have to change this back to return a regular old dict and comment on the typing in the method. What do you think?

Reply:
Yeah, I can see why updating the TypedDict could be type unsafe. What if you `.pop` a required k-v from the dict, now you're passing around a bad TypedDict.

I've swapped out the TypedDict usage with a general Dict and just added a docstring detailing what can be in the return dict.